transformer_engine/pytorch/ops/fuser.py

```diff
@@ -398,27 +398,30 @@ def maybe_fuse_ops(
                     break
 
         # Early exit if fusion parameters haven't changed
+        need_reset = False
         recipe_type = type(recipe)
         fusion_params = (recipe_type, first_op_requiring_backward)
-        if fusion_params == (self.recipe_type, self.first_op_requiring_backward):
+        if fusion_params != (self.recipe_type, self.first_op_requiring_backward):
+            # Recipe type or grad requirmenets have changed
+            need_reset = True
+        elif (
+            recipe is not None
+            and recipe.delayed()
+            and self._last_amax_history_len != recipe.amax_history_len
+        ):
+            # FP8 delayed scaling has changed amax history length
+            need_reset = True
+        if not need_reset:
             return
 
-        # Initialize ops if recipe type has changed
-        if self.recipe_type != recipe_type:
-            # Check if this is the first iteration
-            if self.recipe_type is None:
-                for op in self._basic_ops:
-                    op.pre_first_fuser_forward()
-            # Inform ops that the recipe type has changed
+        # Reset recipe state
+        for op in self._basic_ops:
+            op.reset_recipe_state(recipe=recipe)
+
+        # Check if this is the first iteration
+        if self.recipe_type is None:
             for op in self._basic_ops:
-                op.reset_recipe_type(recipe=recipe)
-        # Check if amax history was invalidated
-        elif isinstance(recipe, DelayedScaling):
-            if recipe.amax_history_len != self._last_amax_history_len:
-                raise RuntimeError(
-                    "Detected change of amax history length. "
-                    "Changing the length of amax history is currently not supported."
-                )
+                op.pre_first_fuser_forward()
```

Comment:
I think these two loops should be reversed (`pre_first_fuser_forward` before `reset_recipe_state`).

Reply:
I think initializing weights with `device="meta"` should be considered an advanced usage, primarily meant for FSDP. If the user initializes quantized weights but does not enable quantization during the first forward pass, I'm fine if that causes an error.

Reply:
Ok, in that case I think the error message in `basic_linear.py:294` could be updated to reflect that. I think something like this could work:
```
                raise RuntimeError(
                    "Tried to quantize weight with deferred initialization "
                    "due to meta device, but no quantizer was available. "
                    "This is most likely because fp8_model_init was called "
                    "with enabled=True, but fp8_autocast was called with "
                    "enabled=False or recipe=None or was not called at all."
                )
```

Reply:
Perhaps the fact that the recipe given to `fp8_model_init` is ignored should also be mentioned somewhere.